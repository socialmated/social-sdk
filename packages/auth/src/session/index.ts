export type * from './session.js';
export * from './oauth.js';
export * from './cookie.js';
