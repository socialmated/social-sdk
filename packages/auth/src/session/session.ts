import { type Promisable } from 'type-fest';

/**
 * Represents a session interface for managing session data, including retrieval,
 * refreshing, revocation, and expiration handling.
 */
export interface ReadonlySession {
  /**
   * Retrieves the value associated with the specified key from the session.
   *
   * @param key - The key whose value should be retrieved.
   * @returns A promise that resolves to the value as a string, or `null` if the key does not exist.
   */
  get: (key: string) => string | null;

  /**
   * Retrieves all key-value pairs stored in the session.
   *
   * @returns A promise that resolves to an object containing all session data.
   */
  getAll: () => Record<string, string>;

  /**
   * Refreshes the session, typically by updating access tokens or other session data.
   *
   * @param key - The key to refresh.
   *
   * @returns A promise that resolves to the refreshed value as a string.
   */
  refresh: (key?: string) => Promisable<string>;

  /**
   * Revokes the session, typically by invalidating access tokens or clearing session data.
   *
   * @returns A promise that resolves when the session is revoked.
   */
  revoke: () => Promisable<void>;

  /**
   * Get the remaining time in seconds until the session expires.
   *
   * @returns The remaining time in seconds until the session expires.
   */
  expiresIn: () => number;

  /**
   * Retrieves the scope of the session.
   *
   * @returns The scope as a string, or `undefined` if not found.
   */
  scope: () => string | undefined;

  /**
   * The issuer of the session, typically the base URL or domain.
   */
  issuer: URL | string;
}

export interface WritableSession extends ReadonlySession {
  /**
   * Sets a value in the session for the specified key.
   *
   * @param key - The key under which the value should be stored.
   * @param value - The value to store in the session.
   */
  set: (key: string, value: string) => void;

  /**
   * Clears the session data, effectively logging out the user or invalidating the session.
   */
  clear: () => void;
}
