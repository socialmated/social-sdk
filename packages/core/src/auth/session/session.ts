/**
 * Represents a session interface for managing session data, including retrieval,
 * refreshing, revocation, and expiration handling.
 */
export interface Session {
  /**
   * Retrieves the value associated with the specified key from the session.
   *
   * @param key - The key whose value should be retrieved.
   * @returns A promise that resolves to the value as a string, or `undefined` if not found.
   */
  get: (key: string) => string | undefined;

  /**
   * Retrieves all key-value pairs stored in the session.
   *
   * @returns A promise that resolves to an object containing all session data.
   */
  getAll: () => Record<string, string>;

  /**
   * Refreshes the session, typically by updating access tokens or other session data.
   *
   * @param key - The key to refresh. If not provided, refreshes all keys.
   *
   * @returns A promise that resolves when the session is refreshed.
   */
  refresh: (key?: string) => Promise<void>;

  /**
   * Revokes the session, typically by invalidating access tokens or clearing session data.
   *
   * @param key - The key to revoke. If not provided, revokes all keys.
   *
   * @returns A promise that resolves when the session is revoked.
   */
  revoke: (key?: string) => Promise<void>;

  /**
   * Get the remaining time in seconds until the session expires.
   *
   * @param key - The key to check for expiration.
   * @returns The remaining time in seconds until the session expires.
   */
  expiresIn: (key?: string) => number;

  /**
   * Retrieves the scope associated with the specified key from the session.
   *
   * @param key - The key whose scope should be retrieved. If not provided, retrieves the default scope.
   * @returns The scope as a string, or `undefined` if not found.
   */
  scope: (key?: string) => string | undefined;

  /**
   * Clears the session data, effectively logging out the user or invalidating the session.
   */
  clear: () => void;

  /**
   * The issuer of the session, typically the base URL or domain.
   */
  issuer: URL | string;
}
