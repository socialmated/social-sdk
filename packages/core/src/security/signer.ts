import { type Options } from 'got';

export interface Signer<T> {
  sign: ((req: Options) => Promise<T>) & ((req: Request) => Promise<T>);
}
