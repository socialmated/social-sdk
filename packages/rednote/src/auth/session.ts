import {
  type CookieJar,
  type LocalStorage,
  type SessionStorage,
  WebStoreCookieSession,
} from '@social-sdk/auth/session';
import { addYear } from '@formkit/tempo';
import { generateLocalId, generateWebId } from '@/security/token/token.js';

export class RednoteCookieSession extends WebStoreCookieSession {
  /**
   * Creates an instance of the session class.
   *
   * @param url - The base URL for the session. Defaults to 'https://xiaohongshu.com' if not provided.
   * @param jar - An optional `CookieJar` instance to manage cookies for the session.
   * @param localStorage - An optional `LocalStorage` instance for managing local storage.
   * @param sessionStorage - An optional `SessionStorage` instance for managing session storage.
   */
  constructor(jar?: CookieJar, localStorage?: LocalStorage, sessionStorage?: SessionStorage) {
    super(new URL('https://www.xiaohongshu.com'), jar, localStorage, sessionStorage);
  }

  /**
   * Wraps a given `WebStoreCookieSession` into a `RednoteCookieSession` instance.
   *
   * @param session - The session object to wrap. Must have an issuer with hostname 'xiaohongshu.com'.
   * @returns A new instance of `RednoteCookieSession` initialized with the provided session's cookie jar, local storage, and session storage.
   * @throws If the session's issuer hostname is not 'xiaohongshu.com'.
   */
  public static wrap(session: WebStoreCookieSession): RednoteCookieSession {
    if (session.issuer.hostname !== 'www.xiaohongshu.com') {
      throw new Error('Invalid session host. Expected www.xiaohongshu.com.');
    }
    return new RednoteCookieSession(session.cookieJar, session.localStorage, session.sessionStorage);
  }

  /**
   * Refreshes and returns a specific session key value.
   *
   * @param key - The session key to refresh ('a1', 'webId', 'loadts', or 'xsecappid')
   * @returns Promise resolving to the refreshed key value
   */
  public override refresh(key: 'a1' | 'webId' | 'loadts' | 'xsecappid'): string {
    if (key === 'a1') {
      const localId = generateLocalId('Mac OS');
      this.set('a1', localId, {
        expires: addYear(new Date(), 1),
        path: '/',
        domain: '.xiaohongshu.com',
      });
      return localId;
    }
    if (key === 'loadts') {
      const ts = String(Date.now());
      this.set('loadts', ts, {
        expires: addYear(new Date(), 1),
        path: '/',
        domain: '.xiaohongshu.com',
      });
      return ts;
    }
    if (key === 'webId') {
      const webId = generateWebId(this.get('a1') ?? generateLocalId('Mac OS'));
      this.set('webId', webId, {
        expires: addYear(new Date(), 1),
        path: '/',
        domain: '.xiaohongshu.com',
      });
      return webId;
    }

    const appId = 'xhs-pc-web';
    this.set('xsecappid', appId, {
      expires: addYear(new Date(), 1),
      path: '/',
      domain: '.xiaohongshu.com',
    });
    return appId;
  }
}
