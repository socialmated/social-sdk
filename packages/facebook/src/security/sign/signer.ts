import { type Signer } from '@social-sdk/client/security';
import { type Options } from '@social-sdk/client/http';
import { generateProof, type AppSecretProof } from './sign.js';

export class AppSecretProofSigner implements Signer<AppSecretProof> {
  constructor(private readonly appSecret: string) {}

  public sign(req: Options): AppSecretProof {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- should be defined in the request
    const accessToken = new URLSearchParams(new URL(req.url!).searchParams).get('access_token');
    if (!accessToken) {
      throw new Error('Access token is required to generate App Secret Proof');
    }

    return generateProof(this.appSecret, accessToken);
  }
}
