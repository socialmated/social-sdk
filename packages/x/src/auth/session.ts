import { CookieSession } from '@social-sdk/core/auth/session';
import { type CookieJar } from 'tough-cookie';
import { addHour, addYear } from '@formkit/tempo';
import { fetchGuestToken, generateCsrfToken } from '../security/token/index.js';

/**
 * Represents a session for the X platform (formerly Twitter) using cookies.
 */
export class XCookieSession extends CookieSession {
  /**
   * Creates an instance of the session class.
   *
   * @param url - The base URL for the session. Defaults to 'https://x.com' if not provided.
   * @param jar - An optional `CookieJar` instance to manage cookies for the session.
   */
  constructor(jar?: CookieJar) {
    super(new URL('https://x.com'), jar);
  }

  /**
   * Wraps a given `CookieSession` into an `XCookieSession` instance.
   *
   * @param session - The `CookieSession` object to be wrapped.
   * @returns An `XCookieSession` instance created from the provided `CookieSession`.
   * @throws If the session's issuer hostname is not `x.com`.
   */
  public static wrap(session: CookieSession): XCookieSession {
    if (session.issuer.hostname !== 'x.com') {
      throw new Error('Invalid session host. Expected x.com.');
    }
    return new XCookieSession(session.cookieJar);
  }

  /**
   * Refreshes the session by updating the guest token.
   *
   * @param key - The key to refresh. Defaults to 'gt' (guest token).
   * @returns A promise that resolves when the guest token has been successfully refreshed.
   */
  public override async refresh(key: 'gt' | 'ct0'): Promise<string> {
    if (key === 'ct0') {
      const csrfToken = generateCsrfToken();
      this.set('ct0', csrfToken, {
        expires: addYear(new Date(), 1),
        path: '/',
        domain: '.x.com',
        secure: true,
        sameSite: 'Lax',
      });
      return csrfToken;
    }

    const guestToken = await fetchGuestToken();
    this.set('gt', guestToken, {
      expires: addHour(new Date(), 2.5),
      path: '/',
      domain: '.x.com',
      secure: true,
      sameSite: 'None',
    });
    return guestToken;
  }
}
