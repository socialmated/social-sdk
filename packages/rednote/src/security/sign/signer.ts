import { type CookieSession } from '@social-sdk/core/auth/session';
import { type Options } from 'got';
import { type Signer } from '@social-sdk/core/security';
import { generateLocalId } from '../token/token.js';
import { signOld, signNew, generateMns, type XhsSignOutput } from './sign.js';
import { defaultConfig } from '@/client/config.js';

/**
 * The output type for the XHS signing process.
 */
interface XhsSignerOptions {
  /**
   * The type of signing to perform.
   */
  type?: 'sign_new' | 'sign_old';

  /**
   * Whether MNS is supported.
   */
  mns?: boolean;
}

/**
 * The `XhsSigner` class provides methods to sign requests for the Xiaohongshu (XHS) API.
 *
 * @example
 * ```typescript
 * const signer = XhsSigner.fromSession(session);
 * const signature = await signer.sign('/api/resource', requestBody, { type: 'sign_new' });
 * ```
 */
class XhsSigner implements Signer<{ 'X-s': string; 'X-t': string; 'X-Mns': string }> {
  /**
   * The `a1` cookie value.
   */
  private a1: string;

  /**
   * Creates an instance of `XhsSigner`.
   * @param a1 - The `a1` cookie value used for signing requests.
   */
  constructor(a1?: string) {
    this.a1 = a1 ?? generateLocalId(defaultConfig.platform);
  }

  /**
   * Creates an instance of `XhsSigner` from a given session.
   *
   * @param session - The `CookieSession` object containing authentication cookies.
   * @returns An instance of `XhsSigner` initialized with the `a1` cookie value from the session.
   * @throws If the session does not contain an `a1` cookie.
   */
  public static fromSession(session: CookieSession): XhsSigner {
    const a1 = session.get('a1');
    if (!a1) {
      throw new Error('Session does not contain a1 cookie');
    }

    return new XhsSigner(a1);
  }

  /**
   * Generates a signature for the given request.
   *
   * @param request - The request options.
   * @param options - Optional parameters for signing, including type and MNS support.
   * @returns An object containing the signed headers, including 'X-s', 'X-t', and 'X-Mns'.
   */
  public sign(request: Options, options?: XhsSignerOptions): XhsSignOutput {
    const fullPath = request.url ? new URL(request.url).pathname + new URL(request.url).search : '';
    const body: unknown =
      !request.json && typeof request.body === 'string' ? JSON.parse(request.body) : (request.json ?? request.body);
    let output: Omit<XhsSignOutput, 'X-Mns'>;

    if (options?.type === 'sign_old') {
      output = signOld(fullPath, body);
    } else {
      output = signNew(fullPath, body, this.a1);
    }

    return {
      ...output,
      'X-Mns': options?.mns ? generateMns(fullPath, body) : 'unload',
    };
  }
}

export { XhsSigner };
export type { XhsSignerOptions };
