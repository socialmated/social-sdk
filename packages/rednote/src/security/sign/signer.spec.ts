import { Options } from 'got';
import { XhsSigner } from './signer.js';
import { RednoteCookieSession } from '@/auth/session.js';

describe(XhsSigner, () => {
  const a1 = '1972697fa3d2ly0o5nfy1iqg0ra0w56q79gz2gm0450000417121';

  let session: RednoteCookieSession;
  let signer: XhsSigner;

  beforeEach(() => {
    vi.setSystemTime(new Date(2025, 0, 1));
    vi.spyOn(Math, 'random').mockReturnValue(0.5);

    session = new RednoteCookieSession();
    signer = new XhsSigner(session);
  });

  describe(XhsSigner.prototype.sign, () => {
    it('should sign POST request with body and return signature', () => {
      session.set('a1', a1);

      const req = new Options({
        method: 'POST',
        url: 'https://edith.xiaohongshu.com/api/sns/web/v1/resource',
        json: {
          key: 'value',
          type: 'example',
        },
      });
      const sig = signer.sign(req);

      expect(sig).toMatchSnapshot();
    });

    it('should sign GET request with query parameters and return signature', () => {
      session.set('a1', a1);

      const req = new Options({
        method: 'GET',
        url: 'https://edith.xiaohongshu.com/api/sns/web/v1/resource',
        searchParams: {
          key: 'value',
          type: 'example',
        },
      });
      const sig = signer.sign(req);

      expect(sig).toMatchSnapshot();
    });

    it('should refresh a1 cookie if not set and sign request', () => {
      const req = new Options({
        method: 'GET',
        url: 'https://edith.xiaohongshu.com/api/sns/web/v1/resource',
      });

      const sig = signer.sign(req);

      expect(sig).toMatchSnapshot();
      expect(session.get('a1')).toBeDefined();
    });
  });
});
