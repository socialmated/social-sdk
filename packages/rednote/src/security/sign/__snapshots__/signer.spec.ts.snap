// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`XhsSigner > sign > should refresh a1 cookie if not set and sign request 1`] = `
{
  "X-Mns": "unload",
  "X-s": "XYW_eyJzaWduU3ZuIjoiNTYiLCJzaWduVHlwZSI6IngyIiwiYXBwSWQiOiJ4aHMtcGMtd2ViIiwic2lnblZlcnNpb24iOiIxIiwicGF5bG9hZCI6IjBiZWU4ZWVmNTQwN2YzYjEyOTU2NDJmNThiNDMyYmI0OWRlMGQ1YjkxOGNmODYyNDc3MzA0NDlkMzEyOGQ0MDE3ZGYyNzY2ZDA0YmMyZGRlNmRkZGQzYWM4MTAyZDMzZjlhYjEzNzE3NWM2NzdlZTIxZjliNTgyNGI2NzcxZjUxZGU2ZGJhZGExOWEzMWZlZjgwNjNhNjJjYzJiY2RmODM0OWIwMGEwMjJjN2I4NmE1NmIzYmE4M2I5NGFjNWY0NTVjMTc1ODFmYzEzY2MzMmMwMDI3MzVjMmJlNzdkZTdkMGFmODFiZDI1NDJiNTI3ODhiODNiNzllODcyY2ZhN2VkNTNmNjVkMTBkNTJhYWViZjNiZDRkZmNkYTIyM2NmYjYzOGM5MDdlNGFhMjc2YjBkZmNmM2U4OTYzNzAxOTVjMDFmMTY0NTJjMTk0OGFkMjJlMzA3YTc0MWZkZTZkNTc0MWJhZjFmNTY0YjY1ZDFlYTQwYmUwMDVlMmJhZGFiNzc2NjEzNmJhYjljNmY0ZWFkYWU3Y2IzMzQyNzc3ODM2In0=",
  "X-t": "1735657200000",
}
`;

exports[`XhsSigner > sign > should sign GET request with query parameters and return signature 1`] = `
{
  "X-Mns": "unload",
  "X-s": "XYW_eyJzaWduU3ZuIjoiNTYiLCJzaWduVHlwZSI6IngyIiwiYXBwSWQiOiJ4aHMtcGMtd2ViIiwic2lnblZlcnNpb24iOiIxIiwicGF5bG9hZCI6IjY2NmZhNTVjZmZkM2Q4NDY3NjVkNzVjN2NhYzUyZjhkMTg3ZTM1MjI3MTkyNDA5NTlmZTZkMjAzNjZkYmYwZDMxZjk0YmFjMzg5NmJiYzRlMmVjYTMzM2ZmMDc5MjQ1Nzk2NTU3YWRjNTFlYmJlMzY3ZGExNzE2NzkxZDdlZTBlYTc5NGNiNWY4NWQ3OTIyY2M4NTg1ODI5YjJhZjRjNjc2YWMxMjVhMGM2ZjQ0Y2M2NTlhMTkzZWQ5MTU4OTZlOTkyNjU3OGI5NjI2NWQ2MjIwMmJkMTM5ZDQ0ZjQ5YjRlZGFhNDNhYzgwNzM2NjI3ZDI3ZDgzOGM0ZGIxODRhZWJlMGFiZWIzNTI0MzVkZjY4NDRjZDU2ZjMzZTUzNGY5YjZjN2M2ZDNjZjMyZTU1OWVjYWVlM2YwOTA1YTYxNWExNzQ5MWNjYjhlOGIwYjJkNWVjYTI5YjI3M2QxZDdkOTYwY2VhYzRhZGZmZWQyMzlhM2JmYTgwNGViZmQ2YzBkZWM5NTNhOTFmYzc2OWYwZDAxOGJhZDc2NWEzYmZmZmFkIn0=",
  "X-t": "1735657200000",
}
`;

exports[`XhsSigner > sign > should sign POST request with body and return signature 1`] = `
{
  "X-Mns": "unload",
  "X-s": "XYW_eyJzaWduU3ZuIjoiNTYiLCJzaWduVHlwZSI6IngyIiwiYXBwSWQiOiJ4aHMtcGMtd2ViIiwic2lnblZlcnNpb24iOiIxIiwicGF5bG9hZCI6IjE2MjEwNTY5YTNjNjFlYzMyZjI3YTI0NWJjNGM2ZDIwYzkxYjhkOGJiOTU1NzMzY2ExOTVhYTJiNWE2MDQ1YWM5YzMzYjc4NDEyNGI2ZmU5M2FhYWY3ZDliZDhlYmIzNTY5YWRlNWFjM2JhZDUxMjYxOWZlMWMxMzYyODgyZTIwYTYwMjZhZjQwMWYyMDM2ZmZlYmVmZTcyMGNhYWRjNTAyMzc2N2QyMDE1NzhkYTI2ZmYzZjhmOGQ4ZDFjNWM3NjY0NTQyNDk3ZTM5ZGFmMjkxN2QyZTUzYjIxYzAxYzUyNDljZTM4NDAyNWFlNzQxY2NjMjRhMjUxNGU4YWQ0ODI5ZmQyOGRmMGJhYjgzNTE2YTMwNTkwYzY4ZTliNWI4ZDA2N2E2OTU0MzJhNTBhNjM3MzIzMzgzMWMxNTNlYTk4OTMxYmYxN2UyYTVhMDBjMTk1ZTkwYjM4YjJiODFhMTIyZDhlOGZmZjNjMWY3NTJiZjk1MTJjY2I5NGM1OTllMTBkMDUxNDNkNTIwODkzMmU1Nzc5ZTFhNmFlNjM1NTEyIn0=",
  "X-t": "1735657200000",
}
`;
